src/dbnode/persist/fs/wide/entry_checksum_mismatch_checker.go

```diff
@@ -0,0 +1,284 @@
+	started   bool
+}
+
+// NewEntryChecksumMismatchChecker creates a new entry checksum mismatch
+// checker, backed by the given block buffer.
+// NB: index entries MUST be checked in lexicographical order by ID.
+func NewEntryChecksumMismatchChecker(
+	blockReader IndexChecksumBlockReader,
+	opts Options,
+) EntryChecksumMismatchChecker {
+	return &entryChecksumMismatchChecker{
+		blockReader: blockReader,
+		mismatches:  make([]ReadMismatch, 0, opts.BatchSize()),
+		decodeOpts:  opts.DecodingOptions(),
+		iOpts:       opts.InstrumentOptions(),
+	}
+}
+
+func entryMismatch(e entryWithChecksum) ReadMismatch {
+	return ReadMismatch{
+		Checksum:    e.idChecksum,
+		EncodedTags: e.entry.EncodedTags,
+		ID:          e.entry.ID,
+	}
+}
+
+func (c *entryChecksumMismatchChecker) entryMismatches(
+	entries ...entryWithChecksum,
+) []ReadMismatch {
+	for _, e := range entries {
+		c.mismatches = append(c.mismatches, entryMismatch(e))
+	}
+
+	return c.mismatches
+}
+
+func (c *entryChecksumMismatchChecker) indexMismatches(checksums ...int64) []ReadMismatch {
+	for _, checksum := range checksums {
+		c.mismatches = append(c.mismatches, ReadMismatch{Checksum: checksum})
+	}
+
+	return c.mismatches
+}
+
+func (c *entryChecksumMismatchChecker) invariant(
+	marker []byte,
+	checksum int64,
+	entry entryWithChecksum,
+) error {
+	// Checksums match but IDs do not. Treat as invariant violation.
+	err := fmt.Errorf("checksum collision")
+	instrument.EmitAndLogInvariantViolation(c.iOpts, func(l *zap.Logger) {
+		l.Error(
+			err.Error(),
+			zap.Int64("checksum", checksum),
+			zap.Binary("marker", marker),
+			zap.Any("entry", entry.entry),
+		)
+	})
+	return err
+}
+
+func (c *entryChecksumMismatchChecker) ComputeMismatchForEntry(
+	indexEntry schema.IndexEntry,
+) ([]ReadMismatch, error) {
+	hash := c.decodeOpts.IndexEntryHasher()
+	checksum := hash.HashIndexEntry(indexEntry)
+	entry := entryWithChecksum{entry: indexEntry, idChecksum: checksum}
+	c.mismatches = c.mismatches[:0]
+	if c.exhausted {
+		// NB: no remaining batches in the index checksum block; any further
+		// elements are mismatches (missing from primary).
+		return c.entryMismatches(entry), nil
+	}
+
+	if !c.started {
+		c.started = true
+		if !c.blockReader.Next() {
+			// NB: no index checksum blocks available; any further
+			// elements are mismatches (missing from primary).
+			c.exhausted = true
+			return c.entryMismatches(entry), nil
+		}
+
+		c.batchIdx = 0
+	}
+
+	batch := c.blockReader.Current()
+	markerIdx := len(batch.Checksums) - 1
+	for {
+		// NB: If the incoming checksum block is empty, move to the next one.
+		if len(batch.Checksums) == 0 {
+			if !c.blockReader.Next() {
+				c.exhausted = true
+				return c.mismatches, nil
+			}
+
+			batch = c.blockReader.Current()
+			markerIdx = len(batch.Checksums) - 1
+			c.batchIdx = 0
+			continue
+		}
+
+		checksum := batch.Checksums[c.batchIdx]
+		compare := bytes.Compare(batch.EndMarker, entry.entry.ID)
+		if c.batchIdx == markerIdx {
```

Comment:
It might help understandability if you would swap this end case and the case with `c.batchIdx < markerIdx` (so that the end is, well, not in the beginning :) ).